### DISCUSSION TITLE
Before/After event for library file functions

### DISCUSSION THREAD

**Discussion by punitkumar** [opencart/opencart#11624] [Q&A]
Is there any way we can call events for library file functions, for example: library file cart.php function getProduct

**Comment by danielkerr** [comment]
it can be done by extending the class and replacing the cart class in the registry via the startup system

**Reply by punitkumar** [reply]
It does not help when there are multiple modules follow same approach. Have look at this 
https://www.loom.com/share/7e0c628ed42b45b2beca93999fa60340

Here the modules archive I used to test
[cart_event_example.zip](https://github.com/opencart/opencart/files/9708956/cart_event_example.zip)

**Reply by mhcwebdesign** [reply]
> It does not help when there are multiple modules follow same approach. Have look at this https://www.loom.com/share/7e0c628ed42b45b2beca93999fa60340
> 
> Here the modules archive I used to test [cart_event_example.zip](https://github.com/opencart/opencart/files/9708956/cart_event_example.zip)

@punitkumar: You are right, that's why we plan to port our Override Engine extension (see https://www.mhccorp.com/opencart/free-items/override-engine-oc3) to OpenCart 4. It will be able to address exactly the issues of different extensions extending the same class such as Cart, and overriding the same method. It use a special factory class and dynamically generates a chain of class extensions.

The only other alternative would be to define before/after triggers for all the methods of a library class, but then you'd end up so quickly with countless trigger points, with the majority of them never been used, and, on the other hand, still potentially wanting more trigger points in other places where they haven't yet been defined. There is only so much you can do with triggers and event handlers.

**Comment by mhcwebdesign** [comment]
If we we want the Cart class to be extendable that way, then this line should be changed:

old:
`private array $data = [];`

new:
`protected array $data = []; `

This way, any overridden methods can make a more meaningful use of the data array, without having to repeat whole DB queries again.

**Reply by punitkumar** [reply]
@mhcwebdesign we not to change access modifier for $data  if we can get return by getProducts() function.
